### PR TITLE
feat(#5665): add shared layout-assertion helpers for browser tests

### LINKED ISSUE #5665: Add shared layout-assertion helpers for browser tests
With Playwright now available in pytest CI (#5658 / #5661), per-issue browser tests can assert layout correctness beyond just "no console errors." I'd like to add a `tests/_layout_helpers.py` module that gives every browser test a one-liner for catching visual layout defects.

## What it provides

- **`collect_layout_violations(page, scope_selector)`** — geometry checks run via `page.evaluate()`:
  - *Overlap*: sibling text-bearing elements whose bounding rects intersect beyond a tolerance (~4px)
  - *Clip*: text content wider than its container without `text-overflow: ellipsis`
  - *Container escape*: element extending past an `overflow: hidden` ancestor
  - *Degenerate box*: visible interactive element with zero width/height or entirely off-viewport
  - *Raw string*: visible text nodes matching `/^[a-z0-9]+(_[a-z0-9]+)+$/` (leaked internal keys surfacing as UI labels)

- **`assert_layout_sane(page, scope_selector)`** — convenience wrapper that calls `collect_layout_violations` and raises `AssertionError` with formatted element paths and rects

- **`assert_no_raw_i18n_keys(page, scope_selector)`** — subset: only the raw-string check

## Motivating defect

The #5554 overlap is the poster child: `.cron-schedule-preset-field{flex:0 0 110px}` with label text wider than 110px and no wrap, so labels collide. A regression test reproducing this with a CSS injection validates the helper catches it.

## Future use

`browser_smoke.py` could adopt the helper later, converting its gate from "no console errors" to "no console errors and no broken layout" for every PR. That's follow-up scope, not this PR.

Nathan approved a shared layout-assertion helper on Discord (2026-07-05).

### PR DESCRIPTION
**Pull request by rodboev**
## Thinking Path

- Playwright is now available in pytest CI after #5661, so per-issue browser tests can run real layout assertions
- Each future test reimplementing geometry checks from scratch would repeat what `test_issue5250` already proved manually, so a shared helper saves that duplication
- The helper runs all checks (overlap, clip, container-escape, degenerate, raw-string) via `page.evaluate()` against any scoped DOM region, and the test validates it against the live app and injected defect fixtures

## What Changed

- `tests/_layout_helpers.py`: new shared helper module exposing `collect_layout_violations(page, scope_selector)`, `assert_layout_sane(page, scope_selector)`, and `assert_no_raw_i18n_keys(page, scope_selector)`. Geometry checks run in-browser via an embedded JS function; Python formats violations.
- `tests/test_layout_helpers.py`: regression tests validating the helper against the live app (all three routes at 1280x720, scoped to `.layout > main` with overlap/clip/container-escape/raw-string checks; degenerate excluded because the composer carries two intentional off-screen a11y proxies) and against injected defect fixtures (flex-row overlap via negative margin, raw i18n key leak).

## Why It Matters

Every future per-issue browser test gets a one-liner for layout correctness. The helper catches the exact class of visual defect that caused maintainer fold-ins on #5554 (flex items overflowing into siblings), and the regression test proves it.

## Verification

```
pytest tests/test_layout_helpers.py -v --timeout=60
pytest tests/test_issue5250_settings_search_dropdown_escape.py -v --timeout=60
```

Full-suite CI context: `pytest tests/ -v --timeout=60`.

## Upstream

Closes #5665.

## Model Used

Claude Opus 4.6 via Claude Code CLI

### CODE REVIEW

File: tests/_layout_helpers.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+function collectRenderViolations(scopeSelector, options) {
+  var opts = options || {}
+  var win = opts.__window || window
+  var doc = opts.__document || win.document || document
+
+  var overlapTol = opts.overlapTolerancePx != null ? opts.overlapTolerancePx : 4
+  var clipTol = opts.clipTolerancePx != null ? opts.clipTolerancePx : 2
+  var escapeTol = opts.escapeTolerancePx != null ? opts.escapeTolerancePx : 2
+  var rawPat = new RegExp(opts.rawStringPattern || "^[a-z0-9]+(_[a-z0-9]+)+$")
+  var enabledChecks = opts.checks || ["overlap", "clip", "container-escape", "degenerate", "raw-string", "a11y"]
```
[the remark this answers is not in the document]
Comment (rodboev) [reply tests/_layout_helpers.py:17]: Intentional. The a11y checks are out of scope for this PR's layout-assertion purpose; the execution prompt explicitly says to omit them from the default checks list. Callers can still enable a11y by passing `checks=[...,'a11y']`.

File: tests/_layout_helpers.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+        }
+      }
+
+      if (enabled("degenerate") && isInteractive(child)) {
+        if (child.tagName === "INPUT" && child.getAttribute("type") === "hidden") {
+          // skip
+        } else if (isVisibleForDegenerate(child)) {
+          var dr = child.getBoundingClientRect()
+          var zeroSize = dr.width === 0 || dr.height === 0
+          var offViewport = dr.right < 0 || dr.bottom < 0 || dr.left > win.innerWidth || dr.top > win.innerHeight
```
[the remark this answers is not in the document]
Comment (rodboev) [reply tests/_layout_helpers.py:229]: This is the canonical JS source embedded verbatim. The boundary condition (strict `>` at the viewport edge) is in the upstream render-lint.js, not introduced by this PR. Changing it here would deviate from the canonical source.

File: tests/_layout_helpers.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    }
+
+    if (enabled("degenerate")) {
+      for (var i = 0; i < degenerateElements.length; i++) {
+        var dNode = degenerateElements[i]
+        if (isInteractive(dNode)) {
+          if (dNode.tagName === "INPUT" && dNode.getAttribute("type") === "hidden") continue
+          var dr2 = dNode.getBoundingClientRect()
+          var zs = dr2.width === 0 || dr2.height === 0
+          var ov = dr2.right < 0 || dr2.bottom < 0 || dr2.left > win.innerWidth || dr2.top > win.innerHeight
```
[the remark this answers is not in the document]
Comment (rodboev) [reply tests/_layout_helpers.py:305]: Same as above: canonical JS source, faithfully embedded. The boundary condition is upstream, not introduced here.

**Comment by nesquena-hermes** [comment]
Shipped in **v0.51.900** (test-only, no production code). Thanks @rodboev — future per-issue browser tests can now use the shared `tests/_layout_helpers.py` layout-lint (overlap / clipping / container-escape / degenerate / raw-i18n-key / a11y) instead of reimplementing geometry checks. Verified: helper imports stdlib only, collection is safe without Playwright installed (tests skip gracefully), and the new tests pass on CI now that #5661 enabled Playwright. Closes #5665.
